README.md
# Ultra Low Power Display



changelog.md
# v1.0.0
- DEV and EPD moved into a library
- debug and dryrun sorta implimented
- usb comms working
- added EPD_pinsDump
- added EPD_pinsCheck
- added EPD_pinsInit
- THE RPI PORTION WORKS

# TODO
- add rotate image
- add mirror image

rush_display.cpp
#include <stdint.h>
#include "pico/stdlib.h"
#include "hardware/spi.h"
#include "EPD.hpp"

#define DEFAULT_WAIT 1*1000
#define LED PICO_DEFAULT_LED_PIN
#define POT_OUT
#define POT_IN
#define POT_ADC
#define LIGHT_OUT
#define LIGHT_IN
#define SCRN_PWR  12
#define SCRN_TX   19
#define SCRN_SCLK 18
#define SCRN_CS   17
#define SCRN_DC   20
#define SCRN_RST  21
#define SCRN_BUSY 22

typedef struct
{
	uint8_t  *image;
	uint16_t width;
	uint16_t height;
	uint8_t  color;
} PAINT;
PAINT paint;

void blink(uint8_t, uint32_t);
void EPD_pinInit(void);

void blink(uint8_t times, uint32_t wait)
{
	for(uint8_t i = 0; i < times; i++)
	{
		gpio_put(LED, 1);
		DEV_delay(wait);
		gpio_put(LED, 0);
		DEV_delay(wait);
	}
}

void EPD_pinInit(void)
{
	EPD_pins.scrn_pwr  = SCRN_PWR;
	EPD_pins.scrn_tx   = SCRN_TX;
	EPD_pins.scrn_sclk = SCRN_SCLK;
	EPD_pins.scrn_cs   = SCRN_CS;
	EPD_pins.scrn_dc   = SCRN_DC;
	EPD_pins.scrn_rst  = SCRN_RST;
	EPD_pins.scrn_busy = SCRN_BUSY;
}

int main()
{
	DEV_delay(3000);
	stdio_init_all();

	gpio_init(LED);
	gpio_set_dir(LED, GPIO_OUT);

	spi_init(spi_default, 1000*1000);
	gpio_set_function(SCRN_SCLK, GPIO_FUNC_SPI);
	gpio_set_function(SCRN_TX, GPIO_FUNC_SPI);
	gpio_set_function(SCRN_CS, GPIO_FUNC_SPI);
	gpio_init(SCRN_PWR);
	gpio_init(SCRN_DC);
	gpio_init(SCRN_RST);
	gpio_init(SCRN_BUSY);
	gpio_set_dir(SCRN_PWR, GPIO_OUT);
	gpio_set_dir(SCRN_DC, GPIO_OUT);
	gpio_set_dir(SCRN_RST, GPIO_OUT);
	gpio_set_dir(SCRN_BUSY, GPIO_OUT);
	EPD_pinInit();

	blink(3, 500);
	printf("Starting board\n");

	while (true) {
		blink(5, 200);
		printf("waiting\n");
		sleep_ms(3000);
	}
}

lib/EPD_board/include/EPD.hpp
#define DRYRUN     1
#define DEBUG      0

#include <stdint.h>
#if DRYRUN || DEBUG
#include <stdio.h>
#endif
#include "DEV.hpp"

#define WHITE      0xFF
#define BLACK      0x00
#define EPD_WIDTH  104
#define EPD_HEIGHT 212

typedef struct
{
	uint8_t scrn_pwr  = 255;
	uint8_t scrn_tx   = 255;
	uint8_t scrn_sclk = 255;
	uint8_t scrn_cs   = 255;
	uint8_t scrn_dc   = 255;
	uint8_t scrn_rst  = 255;
	uint8_t scrn_busy = 255;
} EPD_PINS;
EPD_PINS EPD_pins;

uint8_t EPD_pinCheck(void);
void    EPD_init(void);
void    EPD_reset(void);
void    EPD_sendCommand(const uint8_t*);
void    EPD_sendData(const uint8_t*);
void    EPD_readBusy(void);
void    EPD_turnOnDisplay(void);
void    EPD_clear(void);
void    EPD_display(const uint8_t*, const uint8_t*);
void    EPD_sleep(void);
void    EPD_exit(void);

#if DRYRUN || DEBUG
void EPD_pinsDump(void);
#endif

lib/EPD_board/src/EPD.cpp
#include "EPD.hpp"

uint8_t EPD_pinCheck(void)
{
	#if DRYRUN || DEBUG
	printf("EPD_pinCheck\n");
	#endif

	if(EPD_pins.scrn_pwr  == 255) return 1;
	if(EPD_pins.scrn_tx   == 255) return 2;
	if(EPD_pins.scrn_sclk == 255) return 3;
	if(EPD_pins.scrn_cs   == 255) return 4;
	if(EPD_pins.scrn_dc   == 255) return 5;
	if(EPD_pins.scrn_rst  == 255) return 6;
	if(EPD_pins.scrn_busy == 255) return 7;
	return 0;
}

void EPD_init(void)
{
	#if DRYRUN || DEBUG
	printf("EPD_init\n");
	#endif

	const uint8_t payload[] =
	{
		0x04,
		0x00, 0x0F, 0x89,
		0x61, 0x68, 0x00, 0xD4,
		0x50, 0x77
	};
	uint8_t payloadI = 0;

	EPD_reset();
	DEV_delay(10);
	EPD_readBusy();

	EPD_sendCommand(&payload[payloadI++]);
	EPD_readBusy();

	EPD_sendCommand(&payload[payloadI++]);
	EPD_sendData(&payload[payloadI++]);
	EPD_sendData(&payload[payloadI++]);

	EPD_sendCommand(&payload[payloadI++]);
	EPD_sendData(&payload[payloadI++]);
	EPD_sendData(&payload[payloadI++]);
	EPD_sendData(&payload[payloadI++]);

	EPD_sendCommand(&payload[payloadI++]);
	EPD_sendData(&payload[payloadI]);

	return;
}

void EPD_reset(void)
{
	#if DRYRUN || DEBUG
	printf("EPD_reset\n");
	#endif

	// DEV_write(SCRN_CS, 1);
	DEV_write(EPD_pins.scrn_rst, 1);
	DEV_delay(200);
	DEV_write(EPD_pins.scrn_rst, 0);
	DEV_delay(1);
	DEV_write(EPD_pins.scrn_rst, 1);
	DEV_delay(200);

	return;
}

void EPD_sendCommand(const uint8_t* reg)
{
	#if DRYRUN || DEBUG
	printf("EPD_sendCommand\n");
	#endif

	DEV_write(EPD_pins.scrn_dc, 0);
	DEV_write(EPD_pins.scrn_cs, 0);
	DEV_spiWrite(reg);
	DEV_write(EPD_pins.scrn_cs, 1);

	return;
}

void EPD_sendData(const uint8_t* data)
{
	#if DRYRUN || DEBUG
	printf("EPD_sendData\n");
	#endif

	DEV_write(EPD_pins.scrn_dc, 1);
	DEV_write(EPD_pins.scrn_cs, 0);
	DEV_spiWrite(data);
	DEV_write(EPD_pins.scrn_cs, 1);

	return;
}

void EPD_readBusy(void)
{
	#if DRYRUN || DEBUG
	printf("EPD_readBusy\n");
	#endif

	uint8_t busy = 1;

	do
	{
		DEV_delay(50);
		busy = DEV_read(EPD_pins.scrn_busy);
		busy =! (busy & 0x01);
	} while(busy);

	return;
}

void EPD_turnOnDisplay(void)
{
	#if DRYRUN || DEBUG
	printf("EPD_turnOnDisplay\n");
	#endif

	const uint8_t payload = 0x12;

	EPD_sendCommand(&payload);
	DEV_delay(100);
	EPD_readBusy();

	return;
}

void EPD_clear(void)
{
	#if DRYRUN || DEBUG
	printf("EPD_clear\n");
	#endif

	uint16_t width  = EPD_WIDTH / 8;
	uint16_t height = EPD_HEIGHT;
	const uint8_t payload[] = 
	{
		0x10, WHITE,
		0x13, WHITE
	};
	uint8_t payloadI = 0;


	EPD_sendCommand(&payload[payloadI++]);
	for(uint16_t j = 0; j < height; j++)
	{
		for(uint16_t i = 0; i < width; i++)
		{
			EPD_sendData(&payload[payloadI]);
		}
	}

	payloadI++;

	EPD_sendCommand(&payload[payloadI++]);
	for(uint16_t j = 0; j < height; j++)
	{
		for(uint16_t i = 0; i < width; i++)
		{
			EPD_sendData(&payload[payloadI]);
		}
	}

	EPD_turnOnDisplay();

	return;
}

void EPD_display(const uint8_t* blackImage, const uint8_t* redImage)
{
	#if DRYRUN || DEBUG
	printf("EPD_display\n");
	#endif

	uint16_t width  = EPD_WIDTH / 8;
	uint16_t height = EPD_HEIGHT;
	const uint8_t payload[] = 
	{
		0x10,
		0x13
	};
	uint8_t payloadI = 0;


	EPD_sendCommand(&payload[payloadI++]);
	for(uint16_t j = 0; j < height; j++)
	{
		for(uint16_t i = 0; i < width; i++)
		{
			EPD_sendData(&blackImage[i + j*width]);
		}
	}

	EPD_sendCommand(&payload[payloadI]);
	for(uint16_t j = 0; j < height; j++)
	{
		for(uint16_t i = 0; i < width; i++)
		{
			EPD_sendData(&blackImage[i + j*width]);
		}
	}

	EPD_turnOnDisplay();

	return;
}

void EPD_sleep(void)
{
	#if DRYRUN || DEBUG
	printf("EPD_sleep\n");
	#endif

	const uint8_t payload[] = 
	{
		0x50, 0xF7,
		0x02,
		0x07, 0xA5
	};
	uint8_t payloadI = 0;

	EPD_sendCommand(&payload[payloadI++]);
	EPD_sendData(&payload[payloadI++]);

	EPD_sendCommand(&payload[payloadI++]);
	EPD_readBusy();
	EPD_sendCommand(&payload[payloadI++]);
	EPD_sendData(&payload[payloadI++]);

	return;
}

void EPD_exit(void)
{
	#if DRYRUN || DEBUG
	printf("EPD_exit\n");
	#endif

	DEV_write(EPD_pins.scrn_cs, 0);
	DEV_write(EPD_pins.scrn_pwr, 0);
	DEV_write(EPD_pins.scrn_dc, 0);
	DEV_write(EPD_pins.scrn_rst, 0);

	return;
}

#if DRYRUN || DEBUG
void EPD_pinsDump(void)
{
	printf("\
		scrn_pwr  = %u\n\
		scrn_tx   = %u\n\
		scrn_sclk = %u\n\
		scrn_cs   = %u\n\
		scrn_dc   = %u\n\
		scrn_rst  = %u\n\
		scrn_busy = %u\n",
		EPD_pins.scrn_pwr,
		EPD_pins.scrn_tx,
		EPD_pins.scrn_sclk,
		EPD_pins.scrn_cs,
		EPD_pins.scrn_dc,
		EPD_pins.scrn_rst,
		EPD_pins.scrn_busy
	);

	return;
}
#endif